<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html><head><title>R: An R package for generalized CDH (gCDH) analysis</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
<link rel="stylesheet" type="text/css" href="R.css">
</head><body>
<h1> An R package for generalized CDH (gCDH) analysis
<img class="toplogo" src="../../../doc/html/logo.jpg" alt="[R logo]">
</h1>
<hr>
<div align="center">
<a href="../../../doc/html/packages.html"><img src="../../../doc/html/left.jpg" alt="[Up]" width="30" height="30" border="0"></a>
<a href="../../../doc/html/index.html"><img src="../../../doc/html/up.jpg" alt="[Top]" width="30" height="30" border="0"></a>
</div><h2>Documentation for package &lsquo;CollapsABEL&rsquo; version 0.10.1</h2>

<ul><li><a href="../DESCRIPTION">DESCRIPTION file</a>.</li>
</ul>

<h2>Help Pages</h2>


<p align="center">
<a href="#A">A</a>
<a href="#B">B</a>
<a href="#C">C</a>
<a href="#D">D</a>
<a href="#E">E</a>
<a href="#F">F</a>
<a href="#G">G</a>
<a href="#H">H</a>
<a href="#I">I</a>
<a href="#L">L</a>
<a href="#M">M</a>
<a href="#N">N</a>
<a href="#P">P</a>
<a href="#R">R</a>
<a href="#S">S</a>
<a href="#T">T</a>
<a href="#V">V</a>
<a href="#misc">misc</a>
</p>


<h2><a name="A">-- A --</a></h2>

<table width="100%">
<tr><td width="25%"><a href="alphaNumeric.html">alphaNumeric</a></td>
<td>Alpha-numeric characters</td></tr>
<tr><td width="25%"><a href="asBigMatrix-methods.html">asBigMatrix-method</a></td>
<td>Coerce an R vector/matrix/data.frame into a big.matrix</td></tr>
<tr><td width="25%"><a href="asBigMatrix-methods.html">asBigMatrix-methods</a></td>
<td>Coerce an R vector/matrix/data.frame into a big.matrix</td></tr>
<tr><td width="25%"><a href="assocFilter.html">assocFilter</a></td>
<td>Filter a PlGwas object by the results of a 'plink -assoc' run</td></tr>
</table>

<h2><a name="B">-- B --</a></h2>

<table width="100%">
<tr><td width="25%"><a href="baseName.html">baseName</a></td>
<td>Basename of a FilePath object</td></tr>
<tr><td width="25%"><a href="bedcollr.html">bedcollr</a></td>
<td>Shift bed files</td></tr>
<tr><td width="25%"><a href="bedSizeCorrect.html">bedSizeCorrect</a></td>
<td>Check whether bed file is of correct size</td></tr>
<tr><td width="25%"><a href="bimCorrectTypes.html">bimCorrectTypes</a></td>
<td>Correct types of bim data.frame</td></tr>
<tr><td width="25%"><a href="bim_fam_info.html">bimInfo</a></td>
<td>.bim and .fam file information functions</td></tr>
<tr><td width="25%"><a href="bim_fam_info.html">bim_fam_info</a></td>
<td>.bim and .fam file information functions</td></tr>
<tr><td width="25%"><a href="bin2DescFilename.html">bin2DescFilename</a></td>
<td>Convert a .bin filename to a .desc filename</td></tr>
<tr><td width="25%"><a href="binPhe.html">binPhe</a></td>
<td>Check whether phenotype of a GWAS is binary</td></tr>
<tr><td width="25%"><a href="bmAddCol.html">bmAddCol</a></td>
<td>Add column(s) to an existing big.matrix</td></tr>
<tr><td width="25%"><a href="bmAttachBin.html">bmAttachBin</a></td>
<td>Attach a big.matrix by its bin filename</td></tr>
<tr><td width="25%"><a href="bmConvertFun.html">bmConvertFun</a></td>
<td>Conversion function to use when appending values to a big.matrix</td></tr>
<tr><td width="25%"><a href="bmFilename.html">bmFilename</a></td>
<td>Generate a big.matrix filename (.bin or .desc)</td></tr>
<tr><td width="25%"><a href="bmFilepath.html">bmFilepath</a></td>
<td>Get the big.matrix file path according to GCDH task tag</td></tr>
<tr><td width="25%"><a href="bytesSnp-PlInfo-method.html">bytesSnp-method</a></td>
<td>Get number of bytes used by each SNP.</td></tr>
</table>

<h2><a name="C">-- C --</a></h2>

<table width="100%">
<tr><td width="25%"><a href="changeByMap.html">changeByMap</a></td>
<td>Transform a vector by a mapping</td></tr>
<tr><td width="25%"><a href="charify.html">charify</a></td>
<td>Convert certain columns of a data.frame to character type</td></tr>
<tr><td width="25%"><a href="checkFileExist.html">checkFileExist</a></td>
<td>Stop when any file does not exist</td></tr>
<tr><td width="25%"><a href="chExt.html">chExt</a></td>
<td>Change extension names</td></tr>
<tr><td width="25%"><a href="colClasses.html">colClasses</a></td>
<td>Get classes of columns of a data.frame</td></tr>
<tr><td width="25%"><a href="colCors.html">colCors</a></td>
<td>Correlation coefficient of column-pairs of two data frames</td></tr>
<tr><td width="25%"><a href="collClear.html">collClear</a></td>
<td>Clear up CollapsABEL workspace</td></tr>
<tr><td width="25%"><a href="collenv.html">collenv</a></td>
<td>An environment for storing CollapsABEL package local variables</td></tr>
<tr><td width="25%"><a href="connectSnpPair.html">connectSnpPair</a></td>
<td>Annotate a pair of SNPs in the contrast Manhattan plot</td></tr>
<tr><td width="25%"><a href="contrastData.html">contrastData</a></td>
<td>Prepare data for 'contrastPlot'</td></tr>
<tr><td width="25%"><a href="contrastPlot.html">contrastPlot</a></td>
<td>Produce contrast Manhattan plot</td></tr>
<tr><td width="25%"><a href="correctDesc.html">correctDesc</a></td>
<td>Correct description of big.matrix</td></tr>
<tr><td width="25%"><a href="correctTypes_methods.html">correctTypes-method</a></td>
<td>Convert columns of a data frame to certain types</td></tr>
<tr><td width="25%"><a href="correctTypes_methods.html">correctTypes_methods</a></td>
<td>Convert columns of a data frame to certain types</td></tr>
<tr><td width="25%"><a href="covarNames.html">covarNames</a></td>
<td>Get covariate names of a GWAS</td></tr>
</table>

<h2><a name="D">-- D --</a></h2>

<table width="100%">
<tr><td width="25%"><a href="defmacro.html">defmacro</a></td>
<td>Define "macros" in R</td></tr>
<tr><td width="25%"><a href="desc2BinFilename.html">desc2BinFilename</a></td>
<td>Convert a .desc filename to a .bin filename</td></tr>
<tr><td width="25%"><a href="dir.create2.html">dir.create2</a></td>
<td>Create directory if it does not already exist</td></tr>
<tr><td width="25%"><a href="dirName.html">dirName</a></td>
<td>Directory name of a file path</td></tr>
</table>

<h2><a name="E">-- E --</a></h2>

<table width="100%">
<tr><td width="25%"><a href="eprint.html">eprint</a></td>
<td>Print quoted expression then its value</td></tr>
<tr><td width="25%"><a href="evalFile.html">evalFile</a></td>
<td>Eval R expressions from a file.</td></tr>
</table>

<h2><a name="F">-- F --</a></h2>

<table width="100%">
<tr><td width="25%"><a href="famCorrectTypes.html">famCorrectTypes</a></td>
<td>Correct types of fam data.frame</td></tr>
<tr><td width="25%"><a href="bim_fam_info.html">famInfo</a></td>
<td>.bim and .fam file information functions</td></tr>
<tr><td width="25%"><a href="fidIid.html">fidIid</a></td>
<td>FID and IID columns from fam file</td></tr>
<tr><td width="25%"><a href="file.create2.html">file.create2</a></td>
<td>Create file if it does not already exist</td></tr>
<tr><td width="25%"><a href="filePath.html">filePath</a></td>
<td>Constructor for FilePath class</td></tr>
<tr><td width="25%"><a href="FilePath-class.html">FilePath-class</a></td>
<td>An S4 class to represent a file path</td></tr>
<tr><td width="25%"><a href="fileSize.html">fileSize</a></td>
<td>Get file size</td></tr>
</table>

<h2><a name="G">-- G --</a></h2>

<table width="100%">
<tr><td width="25%"><a href="gcdhBmCreate.html">gcdhBmCreate</a></td>
<td>Create a big.matrix under specified GCDH tag</td></tr>
<tr><td width="25%"><a href="gcdhDir.html">gcdhDir</a></td>
<td>Create GCDH task directories by tag</td></tr>
<tr><td width="25%"><a href="gcdhPower.html">gcdhPower</a></td>
<td>GCDH power analysis</td></tr>
<tr><td width="25%"><a href="gcdhRegion.html">gcdhRegion</a></td>
<td>Run GCDH over a region</td></tr>
<tr><td width="25%"><a href="gcdhReport.html">gcdhReport</a></td>
<td>Generate a report from a GCDH run</td></tr>
<tr><td width="25%"><a href="getOrElse-operator.html">getOrElse-operator</a></td>
<td>Default value for expression.</td></tr>
<tr><td width="25%"><a href="getQuery.html">getQuery</a></td>
<td>Get query results from a SQLite database</td></tr>
<tr><td width="25%"><a href="glm2.html">glm2</a></td>
<td>GLM with arbitrary column names</td></tr>
<tr><td width="25%"><a href="glmIter.html">glmIter</a></td>
<td>Perform glm iteratively through a number of independent variables with fixed dependent variables and covariates.</td></tr>
<tr><td width="25%"><a href="gwasDat.html">gwasDat</a></td>
<td>Read genotype and phenotype data into R</td></tr>
<tr><td width="25%"><a href="gwasDir_methods.html">gwasDir_methods</a></td>
<td>GWAS results directory of a certain GWAS scan</td></tr>
<tr><td width="25%"><a href="gwasFinished.html">gwasFinished</a></td>
<td>Check whether a plink job is finished.</td></tr>
<tr><td width="25%"><a href="gwasLog.html">gwasLog</a></td>
<td>Plink log fie</td></tr>
<tr><td width="25%"><a href="gwasOut.html">gwasOut</a></td>
<td>GWAS output file name</td></tr>
<tr><td width="25%"><a href="gwasOutStem.html">gwasOutStem</a></td>
<td>Plink output filename</td></tr>
<tr><td width="25%"><a href="gwasPercent.html">gwasPercent</a></td>
<td>Check progress of a plink job (by reading its log file)</td></tr>
<tr><td width="25%"><a href="gwasPercentApprox.html">gwasPercentApprox</a></td>
<td>Check progress of a plink job (approximately)</td></tr>
<tr><td width="25%"><a href="gwasR.html">gwasR</a></td>
<td>Invoke a GWAS in R</td></tr>
<tr><td width="25%"><a href="gwasRDS.html">gwasRDS</a></td>
<td>Get RDS file path of a PlGwas object</td></tr>
</table>

<h2><a name="H">-- H --</a></h2>

<table width="100%">
<tr><td width="25%"><a href="head2.html">head2</a></td>
<td>Head and tail in two dimensions</td></tr>
<tr><td width="25%"><a href="headPhe.html">headPhe</a></td>
<td>Read first n lines of a phenotype file</td></tr>
</table>

<h2><a name="I">-- I --</a></h2>

<table width="100%">
<tr><td width="25%"><a href="ifLen.html">ifLen</a></td>
<td>IfLen macro</td></tr>
<tr><td width="25%"><a href="ifLet.html">ifLet</a></td>
<td>IfLet macro</td></tr>
<tr><td width="25%"><a href="ifLetLen.html">ifLetLen</a></td>
<td>IfLetLen macro</td></tr>
<tr><td width="25%"><a href="isBinary.html">isBinary</a></td>
<td>Check whether a trait is binary</td></tr>
<tr><td width="25%"><a href="isS4Class.html">isS4Class</a></td>
<td>Check whether an S4 object is of a certain class</td></tr>
<tr><td width="25%"><a href="isSetup-PlInfo-method.html">isSetup-method</a></td>
<td>Check if a directory containing .bed .fam and .bim files is properly setup</td></tr>
<tr><td width="25%"><a href="isSetupRbed.html">isSetupRbed</a></td>
<td>Check if an RbedInfo object is properly set up</td></tr>
<tr><td width="25%"><a href="isSQLite3.html">isSQLite3</a></td>
<td>Check whether a file is a SQLite3 database.</td></tr>
</table>

<h2><a name="L">-- L --</a></h2>

<table width="100%">
<tr><td width="25%"><a href="lagDistance.html">lagDistance</a></td>
<td>Distance with lag</td></tr>
<tr><td width="25%"><a href="lenCheck.html">lenCheck</a></td>
<td>Check each element of a list has expected length</td></tr>
<tr><td width="25%"><a href="listEqual.html">listEqual</a></td>
<td>Check equality of two lists</td></tr>
<tr><td width="25%"><a href="listGwasTags.html">listGwasTags</a></td>
<td>List GWAS or GCDH tags</td></tr>
<tr><td width="25%"><a href="listGwasTags.html">listTags</a></td>
<td>List GWAS or GCDH tags</td></tr>
<tr><td width="25%"><a href="loadGwas.html">loadGwas</a></td>
<td>Load PlGwas object by tag, from the RDS file</td></tr>
</table>

<h2><a name="M">-- M --</a></h2>

<table width="100%">
<tr><td width="25%"><a href="manhattanData.html">manhattanData</a></td>
<td>Prepare data for Manhattan plot.</td></tr>
<tr><td width="25%"><a href="manhattanPlot.html">manhattanPlot</a></td>
<td>Produce Manhattan plot</td></tr>
</table>

<h2><a name="N">-- N --</a></h2>

<table width="100%">
<tr><td width="25%"><a href="nIndivApprPl-PlInfo-method.html">nIndivApprPl-method</a></td>
<td>Get apparent number of individuals</td></tr>
<tr><td width="25%"><a href="nIndivPl-PlInfo-method.html">nIndivPl-method</a></td>
<td>Get number of individuals</td></tr>
<tr><td width="25%"><a href="nonExistentFiles-character-method.html">nonExistentFiles-method</a></td>
<td>Non-existent files from a vector of filenames</td></tr>
<tr><td width="25%"><a href="nSnpPl-PlInfo-method.html">nSnpPl-method</a></td>
<td>Get number of SNPs.</td></tr>
<tr><td width="25%"><a href="strVectorRepr_methods.html">numVectorSQLRepr</a></td>
<td>String Representation of a character vector</td></tr>
</table>

<h2><a name="P">-- P --</a></h2>

<table width="100%">
<tr><td width="25%"><a href="permutePhe.html">permutePhe</a></td>
<td>Permute a phenotype file</td></tr>
<tr><td width="25%"><a href="PlGwas.html">PlGwas</a></td>
<td>An S4 class representing info about GWAS on plink files</td></tr>
<tr><td width="25%"><a href="plGwas_methods.html">plGwas-method</a></td>
<td>Constructor for PlGwas class</td></tr>
<tr><td width="25%"><a href="plGwas_methods.html">plGwas_methods</a></td>
<td>Constructor for PlGwas class</td></tr>
<tr><td width="25%"><a href="PlInfo.html">PlInfo</a></td>
<td>An S4 class representing info about plink files</td></tr>
<tr><td width="25%"><a href="PlInfo_constructor.html">plInfo-method</a></td>
<td>Constructor for PlInfo class</td></tr>
<tr><td width="25%"><a href="PlInfo_constructor.html">PlInfo_constructor</a></td>
<td>Constructor for PlInfo class</td></tr>
<tr><td width="25%"><a href="plinkr.html">plinkr</a></td>
<td>A wrapper for plink</td></tr>
<tr><td width="25%"><a href="plTrim.html">plTrim</a></td>
<td>Trim plink files</td></tr>
</table>

<h2><a name="R">-- R --</a></h2>

<table width="100%">
<tr><td width="25%"><a href="randNormDat.html">randNormDat</a></td>
<td>Generate a m by n data.frame from normal distribution</td></tr>
<tr><td width="25%"><a href="randomString.html">randomString</a></td>
<td>Generate a single alpha-numeric random string</td></tr>
<tr><td width="25%"><a href="randomStrings.html">randomStrings</a></td>
<td>Generate random strings</td></tr>
<tr><td width="25%"><a href="rbedInfo.html">rbedInfo</a></td>
<td>Constructor of RbedInfo class</td></tr>
<tr><td width="25%"><a href="RbedInfo-class.html">RbedInfo-class</a></td>
<td>S4 class for necessary info to read a bed file into R</td></tr>
<tr><td width="25%"><a href="readAssoc.html">readAssoc</a></td>
<td>Read .qassoc files</td></tr>
<tr><td width="25%"><a href="readBed.html">readBed</a></td>
<td>Read genotypes from PLINK bed file into R</td></tr>
<tr><td width="25%"><a href="readBed.html">readBed-method</a></td>
<td>Read genotypes from PLINK bed file into R</td></tr>
<tr><td width="25%"><a href="readBim.html">readBim</a></td>
<td>Read plink .bim files</td></tr>
<tr><td width="25%"><a href="readBmBin.html">readBmBin</a></td>
<td>Read columns into an R matrix from a big.matrix .bin file</td></tr>
<tr><td width="25%"><a href="readDesc.html">readDesc</a></td>
<td>Read big.matrix .desc file</td></tr>
<tr><td width="25%"><a href="readFam.html">readFam</a></td>
<td>Read plink .fam files</td></tr>
<tr><td width="25%"><a href="readFunFactory.html">readFunFactory</a></td>
<td>Generate read_fun for ReadInfo class</td></tr>
<tr><td width="25%"><a href="readGwasLogLines.html">readGwasLogLines</a></td>
<td>Read stdout from plink into a character vector, each element is a line</td></tr>
<tr><td width="25%"><a href="readGwasLogStr.html">readGwasLogStr</a></td>
<td>Read stdout from plink as a string</td></tr>
<tr><td width="25%"><a href="readGwasOut.html">readGwasOut</a></td>
<td>Read GWAS output from plink If the GWAS is finished, returns a data.frame, otherwise returns NULL.</td></tr>
<tr><td width="25%"><a href="readInfo.html">readInfo</a></td>
<td>ReadInfo constructor</td></tr>
<tr><td width="25%"><a href="ReadInfo-class.html">ReadInfo-class</a></td>
<td>An S4 class to represent information about a whitespace-delimited text file to be read into R</td></tr>
<tr><td width="25%"><a href="readInfo.html">readInfo-method</a></td>
<td>ReadInfo constructor</td></tr>
<tr><td width="25%"><a href="readAssoc.html">readLinear</a></td>
<td>Read .qassoc files</td></tr>
<tr><td width="25%"><a href="readLiteral.html">readLiteral</a></td>
<td>Read a file literally (all columns as character)</td></tr>
<tr><td width="25%"><a href="readAssoc.html">readLogistic</a></td>
<td>Read .qassoc files</td></tr>
<tr><td width="25%"><a href="readPhe.html">readPhe</a></td>
<td>Read phenotype file</td></tr>
<tr><td width="25%"><a href="readPlinkOut.html">readPlinkOut</a></td>
<td>Read plink output files</td></tr>
<tr><td width="25%"><a href="readAssoc.html">readQassoc</a></td>
<td>Read .qassoc files</td></tr>
<tr><td width="25%"><a href="realBedSize.html">realBedSize</a></td>
<td>File size of bed file</td></tr>
<tr><td width="25%"><a href="removeGwasTag.html">removeAllTags</a></td>
<td>Remove GWAS results by tag</td></tr>
<tr><td width="25%"><a href="removeGwasTag.html">removeGwasTag</a></td>
<td>Remove GWAS results by tag</td></tr>
<tr><td width="25%"><a href="removeGwasTag.html">removeTag</a></td>
<td>Remove GWAS results by tag</td></tr>
<tr><td width="25%"><a href="reprClasses.html">reprClasses</a></td>
<td>Represent classes of a data.frame in a character vector</td></tr>
<tr><td width="25%"><a href="rmFilesByStem.html">rmFilesByStem</a></td>
<td>Remove files by matching the starting part</td></tr>
<tr><td width="25%"><a href="runGcdh.html">runGcdh</a></td>
<td>Run GCDH analysis</td></tr>
<tr><td width="25%"><a href="runGwas.html">runGwas</a></td>
<td>Run a GWAS</td></tr>
<tr><td width="25%"><a href="runTypeI.html">runTypeI</a></td>
<td>Run simulations to control type-I error</td></tr>
</table>

<h2><a name="S">-- S --</a></h2>

<table width="100%">
<tr><td width="25%"><a href="saveDesc.html">saveDesc</a></td>
<td>Save big.matrix description object to disk</td></tr>
<tr><td width="25%"><a href="sendQuery.html">sendQuery</a></td>
<td>Send query to SQLite database</td></tr>
<tr><td width="25%"><a href="setOptModel.html">setOptModel</a></td>
<td>Set analysis model</td></tr>
<tr><td width="25%"><a href="setup-PlInfo-method.html">setup-method</a></td>
<td>Setup up a directory containing plink files</td></tr>
<tr><td width="25%"><a href="setupRbed.html">setupRbed</a></td>
<td>Setup an RbedInfo object</td></tr>
<tr><td width="25%"><a href="shiftBed.html">shiftBed</a></td>
<td>Shift bed files</td></tr>
<tr><td width="25%"><a href="shiftedStem.html">shiftedStem</a></td>
<td>Add a "shift" suffix to a stem</td></tr>
<tr><td width="25%"><a href="slurp.html">slurp</a></td>
<td>Read a text file into a single string</td></tr>
<tr><td width="25%"><a href="snpRowId.html">snpRowId</a></td>
<td>Get row number of SNPs from their names</td></tr>
<tr><td width="25%"><a href="spit.html">spit</a></td>
<td>Write strings to a file</td></tr>
<tr><td width="25%"><a href="sqliteFilePl.html">sqliteFilePl</a></td>
<td>SQLite file of a PlInfo object</td></tr>
<tr><td width="25%"><a href="stopFormat.html">stopFormat</a></td>
<td>Stop with format string</td></tr>
<tr><td width="25%"><a href="strConcat.html">strConcat</a></td>
<td>Concatenate a vector of strings</td></tr>
<tr><td width="25%"><a href="strConcat.html">strConcat-method</a></td>
<td>Concatenate a vector of strings</td></tr>
<tr><td width="25%"><a href="strVectorRepr_methods.html">strVectorRepr-method</a></td>
<td>String Representation of a character vector</td></tr>
<tr><td width="25%"><a href="strVectorRepr_methods.html">strVectorRepr_methods</a></td>
<td>String Representation of a character vector</td></tr>
<tr><td width="25%"><a href="strVectorRepr_methods.html">strVectorSQLRepr</a></td>
<td>String Representation of a character vector</td></tr>
<tr><td width="25%"><a href="systemFormat.html">systemFormat</a></td>
<td>Call system command with format string</td></tr>
</table>

<h2><a name="T">-- T --</a></h2>

<table width="100%">
<tr><td width="25%"><a href="head2.html">tail2</a></td>
<td>Head and tail in two dimensions</td></tr>
<tr><td width="25%"><a href="theoBedSize.html">theoBedSize</a></td>
<td>Theoretical size of bed file</td></tr>
<tr><td width="25%"><a href="theoPlinkOutSize.html">theoPlinkOutSize</a></td>
<td>Approximate output file size of a plink job.</td></tr>
</table>

<h2><a name="V">-- V --</a></h2>

<table width="100%">
<tr><td width="25%"><a href="validPhe.html">validPhe</a></td>
<td>Validate a phenotype file</td></tr>
</table>

<h2><a name="misc">-- misc --</a></h2>

<table width="100%">
<tr><td width="25%"><a href="getOrElse-operator.html">%||%</a></td>
<td>Default value for expression.</td></tr>
<tr><td width="25%"><a href="FilePath-class.html">.FilePath</a></td>
<td>An S4 class to represent a file path</td></tr>
<tr><td width="25%"><a href="PlGwas.html">.PlGwas</a></td>
<td>An S4 class representing info about GWAS on plink files</td></tr>
<tr><td width="25%"><a href="PlInfo.html">.PlInfo</a></td>
<td>An S4 class representing info about plink files</td></tr>
<tr><td width="25%"><a href="RbedInfo-class.html">.RbedInfo</a></td>
<td>S4 class for necessary info to read a bed file into R</td></tr>
<tr><td width="25%"><a href="ReadInfo-class.html">.ReadInfo</a></td>
<td>An S4 class to represent information about a whitespace-delimited text file to be read into R</td></tr>
</table>
</body></html>
